_source/_posts/2019-12-20-five-tools-improve-java.adoc

```diff
@@ -0,0 +1,345 @@
+        </execution>
+
+        <!-- generate an HTML report -->
+        <execution>
+            <id>report</id>
+            <phase>verify</phase>
+            <goals>
+                <goal>report</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+
+Run `mvn verify` to generate the report in `target/site/jacoco/index.html`.
+
+image:{% asset_path 'blog/five-tools-improve-java/jacoco-package.png' %}[alt=JaCoCo package screen shot ...,,align=center]
+
+Above is a screenshot of the test coverage report for this project. You can see right away that the `FiltUtility` class has no coverage and that it has more lines than `SomeUtilityClass`.  If click into `SomeUtilityClass`, you can see the tests have missed one of the branches on line 8.
+
+image:{% asset_path 'blog/five-tools-improve-java/jacoco-class.png' %}[alt=JaCoCo class screen shot ...,,align=center]
+
+If you have a Maven multi-module project, you can https://github.com/jacoco/jacoco/wiki/MavenMultiModule[aggregate coverage data] across modules and even merge unit test and integration test coverage to get a picture of the full project's coverage.
+
+== 2. Do Static Source Code Analysis with PMD
+
+PMD is a static code analyzer that can detect potential issues such as dead code, empty blocks, complicated statements, suboptimal code, and duplicate code. PMD doesn't stand for anything; unofficially, some refer to it as "Programming Mistake Detector."
+
+image:{% asset_path 'blog/five-tools-improve-java/pmd-logo.png' %}[alt=PMD logo ...,,align=center]
+
+PMD's Maven plugin one of the officially supported plugins in the Apache Maven project and comes with a robust set of default rules; however, if you are not happy with these rules, you can change them or https://pmd.github.io/pmd/pmd_userdocs_making_rulesets.html[define your own]. The basic configuration for a `pom.xml` is as follows:
+
+.pom.xml
+[source,xml]
+----
+<plugin>
+    <groupId>org.apache.maven.plugins</groupId>
+    <artifactId>maven-pmd-plugin</artifactId>
+    <version>3.12.0</version>
+    <executions>
+        <execution>
+            <id>pmd-scan</id>
+            <phase>verify</phase>
+            <goals>
+                <goal>check</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+
+Now run `mvn verify` or just `mvn pmd:check` on a project with the following Java class:
+
+.FileUtility.java
+[source,java]
+----
+package com.okta.examples;
+
+import java.io.File;
+import java.io.FileWriter;; // <1>
+import java.io.IOException;
+import java.io.BufferedOutputStream; // <2>
+
+public class FileUtility {
+
+    private static void toFile_bad(String contents, File file) throws IOException { // <3>
+        FileWriter fileWriter = new FileWriter(file);
+        fileWriter.write(contents);
+    }
+}
+----
+<1> Empty statement `;;` - delete the trailing `;`.
+<2> Unused import - remove line.
+<3> Unused private method - dead code, consider deleting.
```

Comment:
If you want this block to look better, add `====` before `[source...]` and after the last number.